Fix #3059: Create test for image view binding adapters

yash10019coder opened the pull request:
<!-- READ ME FIRST: Please fill in the explanation section below and check off every point from the Essential Checklist! -->
## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #3059

Added tests for bindable adapters in ImageViewBindingAdapters
## Essential Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

## For UI-specific PRs only
<!-- Delete these section if this PR does not include UI-related changes. -->
If your PR includes UI-related changes, then:
- Add screenshots for portrait/landscape for both a tablet & phone of the before & after UI changes
- For the screenshots above, include both English and pseudo-localized (RTL) screenshots (see [RTL guide](https://github.com/oppia/oppia-android/wiki/RTL-Guidelines))
- Add a video showing the full UX flow with a screen reader enabled (see [accessibility guide](https://github.com/oppia/oppia-android/wiki/Accessibility-(A11y)-Guide))
- Add a screenshot demonstrating that you ran affected Espresso tests locally & that they're passing

yash10019coder committed:
added test activity file and writeen some starter code on it

yash10019coder committed:
commit for test cases

yash10019coder committed:
Added test activity for ImageViewBindingAdapters

yash10019coder committed:
added before and after in test file

yash10019coder committed:
added first test case

yash10019coder committed:
modified file as static checks were failing

yash10019coder committed:
modified file as static checks were failing

yash10019coder committed:
added comment in ImageViewBindingAdaptersTestActivity

yash10019coder committed:
added runConfiguration.xml file

yash10019coder committed:
added runConfiguration.xml file

yash10019coder committed:
Added runConfigurations.xml and also made the activity injectable

yash10019coder committed:
Added imports lexiographically and also removed unused imports

yash10019coder committed:
Added custom matcher for drawable

yash10019coder committed:
removed unused imports

yash10019coder committed:
sorted in lexiograhic order

yash10019coder committed:
added the line in gradle.properties

yash10019coder committed:
added the drawable matcher form Espresso test matchers

yash10019coder committed:
Revert "added the line in gradle.properties"

This reverts commit 98cd838358f3c2913483abae3a6f2418a2b67f2c.

yash10019coder committed:
added the drawable matcher form Espresso test matchers and also removed a obsolete file

yash10019coder committed:
added test case for glide method

yash10019coder committed:
formatted the file

yash10019coder committed:
Added a test for glide image drawable

yash10019coder committed:
Added a test for glide image drawable

yash10019coder committed:
added test for static ImageViewBindingAdapter

yash10019coder committed:
removed unused imports

yash10019coder committed:
Merge branch 'develop' into create-test-for-image-view-binding-adapters

yash10019coder committed:
Added inject function for test activity

yash10019coder commented:
Hi @rt4914 getting this error please help
![image](https://user-images.githubusercontent.com/76404817/138221658-70b1a05a-88db-4870-b559-35e8e5cf4c9f.png)
![image](https://user-images.githubusercontent.com/76404817/138221748-24fd4ceb-c7df-4dde-a0a2-74019f1b7c1c.png)

BenHenning commented:
@yash10019coder this is another case where you should create a debugging doc to help provide context. In particular, it would be good to understand what you think the issue is, and how you think it should be fixed (that way you can also get feedback on how you're trying to investigate the issue, not just the issue itself).

yash10019coder committed:
Merge branch 'develop' into create-test-for-image-view-binding-adapters

yash10019coder committed:
merging develop in this branch

BenHenning commented:
@yash10019coder did you mean to close this?

yash10019coder committed:
Merge branch 'oppia:develop' into create-test-for-image-view-binding-adapters

yash10019coder committed:
Merge branch 'create-test-for-image-view-binding-adapters' of github.com:yash10019coder/oppia-android into create-test-for-image-view-binding-adapters

yash10019coder committed:
removed multiple activities in app/AndroidManifest.xml

yash10019coder committed:
removed multiple activities in ActivityComponentImpl.kt.

yash10019coder committed:
added tests for first two custom adapters for ImageViewBindingAdapters in ImageViewBindingAdaptersTest.kt

yash10019coder committed:
removed unused imports.

yash10019coder committed:
Merge branch 'oppia:develop' into create-test-for-image-view-binding-adapters

yash10019coder committed:
Merge branch 'oppia:develop' into create-test-for-image-view-binding-adapters

yash10019coder committed:
added test.

yash10019coder committed:
Merge branch 'create-test-for-image-view-binding-adapters' of github.com:yash10019coder/oppia-android into create-test-for-image-view-binding-adapters

yash10019coder committed:
added couroutines dispatcher.

yash10019coder committed:
nit fixes.

BenHenning commented:
Hi. As of today, some main reviewers have taken time off for the next few weeks, so it may take a little while before we can look at this PR. We appreciate your patience while some of our team members recharge. We'll be fully returning on 4 January 2021.

yash10019coder committed:
added a url.

yash10019coder committed:
removed the obsolete function.

yash10019coder committed:
Merge branch 'develop' of github.com:oppia/oppia-android into create-test-for-image-view-binding-adapters

yash10019coder committed:
added the injector and injector provider for ImageLoader
and also added static injectors in ImageViewBindingAdapters.java

yash10019coder committed:
applied injectors in OppiaApplication.kt and other places.

yash10019coder reviewed (commented):
Hi @BenHenning I have made the changes you requested but on the running the test build is failing [here's the gist link](https://gist.github.com/410f94dca33434a2cf258d9ef98b87a9) PTAL thanks

rt4914 commented:
Will have to review it once its approve by @BenHenning as he has more context about this compared to me.

BenHenning commented:
@yash10019coder the way you're setting up the injector looks correct, but you'll also need to update every failing test's TestApplicationComponent in the same way you did in the tests here to ensure the injector is set up for those (otherwise they won't build).

Further, you'll also need to update the instrumentation version of OppiaApplication in the same way you updated OppiaApplication.

Please assign back once CI workflows are passing.

yash10019coder committed:
fixed dagger injections.

yash10019coder committed:
added ImageLoaderModule.kt

yash10019coder reviewed (commented):
@BenHenning I fixed all the dagger related problems but now as I created the module for ImageLoader and added into the components of ImageViewBindingAdaptersTest I get a new error [link](https://gist.github.com/11c7404054f985d3656c082b29dc89e0) I get this please help PTAL thanks

yash10019coder committed:
removed ImageLoaderModule.kt as it is already present
and nit fixes.

yash10019coder committed:
nit fixes.

yash10019coder committed:
added injector and injector provider in test_file_exemptions.textproto

yash10019coder committed:
Merge branch 'develop' of github.com:oppia/oppia-android into create-test-for-image-view-binding-adapters

yash10019coder committed:
added bazel targets.

yash10019coder committed:
removed unused methods.

BenHenning commented:
Apologies for the late follow-up; I haven't caught up on all my emails from the last few days yet.

The error log is indicating a duplicate binding. For some reason, both TestImageLoaderModule and ImageLoaderModule are being pulled into the build graph for ImageViewBindingAdaptersTest which shouldn't be happening. I'm actually not 100% sure why this is happening. I need to think on this.

yash10019coder committed:
nit fixes.

yash10019coder committed:
removed unnecessary changes nit formatting fixes.

yash10019coder commented on app/src/sharedTest/java/org/oppia/android/app/databinding/ImageViewBindingAdaptersTest.kt:
```diff
@@ -111,11 +138,78 @@ class ImageViewBindingAdaptersTest {
+      setProfileImage(imageView, profileAvatar)
+      testCoroutineDispatchers.runCurrent()
+      lateinit var imageBitmapSource: ImageTarget<Bitmap>
+      testGlideImageLoader.loadBitmap(url, imageBitmapSource)
```
Am I using this loadBitmap function correctly I'm having doubt in this.

yash10019coder commented on app/src/sharedTest/java/org/oppia/android/app/databinding/ImageViewBindingAdaptersTest.kt:
```diff
@@ -111,11 +138,78 @@ class ImageViewBindingAdaptersTest {
+      val url = profileAvatar.avatarImageUri.toString()
+      setProfileImage(imageView, profileAvatar)
+      testCoroutineDispatchers.runCurrent()
+      lateinit var imageBitmapSource: ImageTarget<Bitmap>
```
This variable remains uninitialized and throws error

BenHenning replied:
You're never assigning this to a value--that's why it's uninitialized.

What, specifically, are you trying to accomplish here? This looks quite different from other tests in the codebase, so I'm wondering if you're trying to do something specific that could be solved with patterns used in other tests. Perhaps look for a similar test elsewhere and model this after one of those.

yash10019coder replied:
I have changed the implementation.
Earlier I thought that loadBitmap(param1,param2)
param1 is the url and param2 is the target variable so I made the variable lateinit and the next line was supposed to initialize it but it never happened.
So now I'm creating a dynamic image view in which i'm loading the image with loadBitmap and then I'll access the BitmapDrawable and pass it to the test function.

yash10019coder reviewed (commented):
Hi @BenHenning the previous error is resolved, but I'm  receiving some new errors PTAL help
 thanks
Here's the link for the gist [link](https://gist.github.com/e1672120e1def1c8641cce6538311577)

and when I try to build locally my build is faling whereas on the github actions it is successfully building 
here's the gist [link](https://gist.github.com/06fdfd936487bb542cda2fd0128ef8ce)

yash10019coder committed:
added bazel deps.

BenHenning reviewed (commented):
> Hi @BenHenning the previous error is resolved, but I'm receiving some new errors PTAL help thanks Here's the link for the gist [link](https://gist.github.com/e1672120e1def1c8641cce6538311577)
> 
> and when I try to build locally my build is faling whereas on the github actions it is successfully building here's the gist [link](https://gist.github.com/06fdfd936487bb542cda2fd0128ef8ce)

Apologies for the delayed reivew here @yash10019coder. It looks like it's failing to run, not build, per the Gist. This is also happening on actions: https://github.com/oppia/oppia-android/runs/4995215171?check_suite_focus=true#step:17:551, so it seems like both your local environment and CI are behaving consistently. I left an additional comment in reply to your comment on the lateinit var with an explanation of what I think is going wrong here.

BenHenning commented:
Also @yash10019coder this doesn't seem to be fixing #2658 (I see no BindableAdapter changes in the PR). Did you mean to add it to this PR's title/description?

yash10019coder commented:
> Also @yash10019coder this doesn't seem to be fixing #2658 (I see no BindableAdapter changes in the PR). Did you mean to add it to this PR's title/description?

My apologies @BenHenning  I got confused I added ImageLoader to dagger graph and was using it in binding adapters
And Shall I add the BindableAdapters to the dagger graph as I have added the ImageLoader?

yash10019coder committed:
used a new implementation for getting bitmap from source.

yash10019coder committed:
nit formatting fixes.

yash10019coder committed:
nit formatting fixes.

yash10019coder committed:
removed unnecessary changes.

yash10019coder reviewed (commented):
Hi @BenHenning addressed all the comments PTAL thanks

yash10019coder commented:
@BenHenning I have changed the implementation but still I get NullPointerException [link for the gist](https://gist.github.com/b7b65837589410ae2094af7393123806)

yash10019coder committed:
added couroutine dispatchers

yash10019coder committed:
Merge branch 'develop' of github.com:oppia/oppia-android into create-test-for-image-view-binding-adapters

yash10019coder committed:
tried to test the matcher.

yash10019coder committed:
nit fixes.

yash10019coder committed:
nit fixes.

BenHenning commented on app/src/sharedTest/java/org/oppia/android/app/databinding/ImageViewBindingAdaptersTest.kt:
```diff
@@ -200,14 +198,13 @@ class ImageViewBindingAdaptersTest {
-      testGlideImageLoader.loadBitmap(url, imageBitmapSource)
+      var imageViewBitmap = ImageView(it)
+      testCoroutineDispatchers.runCurrent()
+      testGlideImageLoader.loadBitmap(url, ImageViewTarget(imageViewBitmap))
```
So before getting into the test, you'll need to update the adapters class to retrieve ImageLoader from the application context (similar to OppiaClockInjector in TextViewBindingAdapters; the principle is the same).

Then you'll need to replace the Glide code with a call to load the image via ImageLoader, and verify manually that this is working. Once verified, you can use TestGlideImageLoader to retrieve the loaded images (you won't need to call loadBitmap here since the adapter method itself will).

BenHenning reviewed (commented):
Thanks @yash10019coder. Left a follow-up comment detailing the parts that I think are still missing here--PTAL.

yash10019coder committed:
Merge branch 'oppia:develop' into create-test-for-image-view-binding-adapters

yash10019coder commented:
Hi @BenHenning are you saying that to change the implementation of the ImageViewBindingAdapters to use the TestGlideImageLoader?

BenHenning commented:
> Hi @BenHenning are you saying that to change the implementation of the ImageViewBindingAdapters to use the TestGlideImageLoader?

Sorry @yash10019coder, I've had a high priority issue that I've been dealing with for the last week.

I'm more or less saying that, but it'd be more correct to say that you need to update ``ImageViewBindingAdapters`` to use ``ImageLoader`` (which, for tests, will be the test image loader and for production code will be the regular Glide image loader). Does that help clarify?
